### ISSUE TITLE
Request: Expose "AllowNewHardware" settings to be used in python plugins

### ISSUE THREAD

**Issue by enesbcs** [domoticz/domoticz#4704]
"AcceptNewHardware" settings value can be seen in JSON and from python plugins:

https://www.domoticz.com/wiki/Domoticz_API/JSON_URL's#Get_settings

But if anyone enables Allow for 5 minutes, "AcceptNewHardware" value is still 0 and there are no other variable for example "AllowNewHardware5" to see if it is enabled or not.

**Comment by rwaaren** [comment]
Can you please describe your usecase for this request?

If you enable this from your plugin with json.htm?type=command&param=allownewhardware&timeout=< time in minutes > and keep note of the start time you know when it will return to the original state.

**Comment by enesbcs** [comment]
User changes this setting in Domoticz GUI.
I've just check its value from inside a python plugin.
https://github.com/enesbcs/Shelly_MQTT/issues/84
https://github.com/enesbcs/Shelly_MQTT/issues/88

**Comment by pipiche38** [comment]
The Plugin Parameter dict is the proper way to get inside a python plugin the various domoticz settings.
Indeed so far the plugin doesn't have info on AcceptnewHardware. It would be good that a notification is send to plugin when enable and when disabled

**Comment by rwaaren** [comment]
@gizmocuz ; please advise.  Should this be exposed and if so how ( separate JSON Call, part of /json.htm?type=settings or something else?)

**Comment by gizmocuz** [comment]
If a plugin wants to do this, I suggest to call 'json.htm?type=command&param=allownewhardware&timeout=5'
But letting the user decide this would be my preference

**Comment by pipiche38** [comment]
@gizmocuz , json.html will require to do an http call , which is not the best in the Python Plugin framework, as it would create a blocking IO and which can create issues in the plugin execution if this call is made inside the main thread (which is for most of the plugin the case ).

For reference: https://www.domoticz.com/wiki/Developing_a_Python_plugin
The following things should not be attempted using the Python Framework:

    Waiting or sleeping in Domoticz callbacks. Plugin callbacks are single threaded so the whole plugin system will wait.

The alternative would be to start an extra thread to do such call (which is linked to more recent development available since 2020.1 )

while I think that if we make the settings available through the existing data structure in place ( Parameter ) it will be easier.

**Comment by gizmocuz** [comment]
It is not a setting, it is a request/action that has to be called, in this case, allow new hardware for 5 minutes
For the end user it would be not a good thing if a plugin can change a setting ... specially if a user does not want to allow new devices.
When adding new hardware, allow for 5 minutes is already enabled I believe

**Comment by pipiche38** [comment]
My suggestion is to do it vice-versa and not having the plugin requesting AcceptingNewHardware, but more using Domoticz to do so and getting domoticz sending a notification to the plugin framework

This would require one more callback which would be trigger by Domoticz every time the AcceptNewHardware is set, it would give as parameters status and the duration ( status 0: disable ;  1, enable ).

This might be implemented like the onNotificationCallback which is already implemented.

**Comment by enesbcs** [comment]
> It is not a setting, it is a request/action that has to be called, in this case, allow new hardware for 5 minutes

I see, but i have a feeling that Domoticz has a variable somewhere inside which contains if it is currently activated.

> For the end user it would be not a good thing if a plugin can change a setting ... specially if a user does not want to allow new devices.

I do not want to change this setting. I want to know if it is in enabled state - from a python plugin - which is a Hardware in this case.

**Comment by enesbcs** [comment]
> @gizmocuz ; please advise. Should this be exposed and if so how ( separate JSON Call, part of /json.htm?type=settings or something else?)

It would be nice if this json call return this state, and i hope that Python Settings[] array will also contain it to let know its current state.

**Comment by kiddigital** [comment]
The settings call already returns 'AcceptNewHardware' (0 or 1)

**Comment by rwaaren** [comment]
> The settings call already returns 'AcceptNewHardware' (0 or 1)


@kiddigital  Yes but that call returns the setting in the database and is not influenced by a call to 

`json.htm?type=command&param=allownewhardware&timeout=5`

**Comment by kiddigital** [comment]
> > The settings call already returns 'AcceptNewHardware' (0 or 1)
> 
> 
> 
> 
> 
> @kiddigital  Yes but that call returns the setting in the database and is not influenced by a call to 
> 
> 
> 
> `json.htm?type=command&param=allownewhardware&timeout=5`
> 
> 
> 
> 

Hmm.. did a quick look in the code and indeed, it only updates the in-memory runtime variables and not the database :( So the log mentions that new devices are allowed for X minutes but the settings screen and database tell something different.

Needs fixing?

**Comment by gizmocuz** [comment]
Yes and it is intended this way

**Comment by gizmocuz** [comment]
A plugin should not be allowed to change application settings (or read critical settings)

**Comment by kiddigital** [comment]
> A plugin should not be allowed to change application settings (or read critical settings)

But having the callable Cmd_AllowNewHardware suggests otherwise.. it is callable (admin only) and now the UI/db/settings json is not in sync with the actual behavior.

The user 'approved' the plugin to use admin privileges (is that actually the case? Can a plugin run as another (non admin) user?), so consequently the plugin can call this...

Maybe at least than report the in-memory state (additionally)?

Or clarify/explain the deliberate 'inconsistency'?

**Comment by enesbcs** [comment]
> A plugin should not be allowed to change application settings (or read critical settings)

Python plugins listed as "Hardware". Why they can not read any setting? Do C++ based hardwares have elevated privilages in front of python hardwares?

**Comment by gizmocuz** [comment]
Yes that is correct, a Admin can change settings, and only the admin is allowed to do so.
Plugings should not be allowed to perform admin related stuff, it should only be allowed to change it's own settings, and create update and delete its own devices.

@enesbcs , Yes, as the source code is stored inside the Domoticz main repository and we can see what it does.
Who knows what a external uncontrolled python plugin can do for harm.
Many years I have already mentioned to create a repository for all plugins here with one centralized manager.

I see no harm in being able to reading the status of **CSQLHelper::m_bAcceptNewHardware** to warn users.
Could also be annoying... for instance if a plugin creates 3 devices, and the user only wants 2, to warn him every minute is not desired

**Comment by enesbcs** [comment]
> @enesbcs , Yes, as the source code is stored inside the Domoticz main repository and we can see what it does.
> Who knows what a external uncontrolled python plugin can do for harm.
> Many years I have already mentioned to create a repository for all plugins here with one centralized manager.

I agree, this is a very good idea!

> I see no harm in being able to reading the status of **CSQLHelper::m_bAcceptNewHardware** to warn users.

Reading "AcceptNewHardware" may be enough, if this variable follows it realtime if the user enables it for 5 minutes. Currently there are no way that a python plugin can told reliably if new hardware creation is enabled or not at the moment.

**Comment by pipiche38** [comment]
@enesbcs my C++ coding is limited, but I know pretty much what the Python Framework provided.  what I have suggested is to have 1 or 2 additional apis in the Framework

The first API would send a notification the python plugins indicating that AcceptNewHardware has been enabled
The second API would consist into quering the status of AcceptNewHardware ( currently the framework is using this call m_sql.m_bAcceptNewHardware to know if a Device can be added or not )


The first on as I mentioned earlier can probably be implemented on the same model as the onNotification call back
The second one seems more complex as I don't think that we have any API like that except the Device one which give the nValue and sValue of one particular unit.

**Comment by cclauss** [comment]
Can Connections, Devices, Images, Parameters, and Settings all be used as Python dictionaries?
https://www.domoticz.com/wiki/Developing_a_Python_plugin#Runtime_Structure says 4 dictionaries but then it lists 5 datatypes.
